core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -462,20 +465,22 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
 
       if (!inPlaceAssignment) {
         // Cannot do in place assignment.
-        val wrapperMessageTimestamp = {
+        val (wrapperMessageTimestamp, offsetOfMaxTimestampInMessageSet) = {
           if (messageFormatVersion == Message.MagicValue_V0)
-            Some(Message.NoTimestamp)
-          else if (messageFormatVersion > Message.MagicValue_V0 && messageTimestampType == TimestampType.CREATE_TIME)
-            Some(maxTimestamp)
+            (Some(Message.NoTimestamp), -1L)
+          else if (messageTimestampType == TimestampType.CREATE_TIME)
+            (Some(maxTimestamp), {if (targetCodec == NoCompressionCodec) offsetOfMaxTimestamp else offsetCounter.value + validatedMessages.length - 1})
           else // Log append time
-            Some(now)
+            (Some(now), {if (targetCodec == NoCompressionCodec) offsetCounter.value else offsetCounter.value + validatedMessages.length - 1})
```

Comment:
Hmm, if the message is not compressed, the offset should be the exact offset of the message that has the largest timestamp. 

When using CREATE_TIME (line 472), that message would be offsetOfMaxTimestamp. But when using LOG_APPEND_TIME, it seems the offset should be the offset of the first message in the message set because all the messages in the validatedMessages will have the same timestamp. (On line 474, offsetCounter has not been incremented yet. So offsetCounter.value is still the offset of first message).

I thought about making wrapperMessageTimestamp non-Option, but we actually will pass in None in the constructor in line 272.